### PR TITLE
Use SerDe to save sierra program

### LINKED ISSUE #510: Replace textual Sierra representation with JSON serialization of the Sierra struct
Currently, when building a `[lib]` target we rely on sierra textual representation. This representation can then be used as an intermediate artifact, e.g. for `cairo-run` extension. This imposes a need to rely on the textual representation parser which is deemed to not be production grade, which may induce hard to trace errors. Instead we should serialize the sierra program to some serde format, that can serve as a "_runnable_" representation.

### PR DESCRIPTION
**Pull request by MrDenkoV**
Fix #510

### CODE REVIEW

File: extensions/scarb-cairo-run/src/main.rs [hunk 1]
```diff
@@ -61,13 +62,13 @@ fn main() -> Result<()> {
...
-            &fs::read_to_string(path.clone())
-                .with_context(|| format!("failed to read Sierra file: {path}"))?,
-        )
-        .map_err(|e| anyhow!("{e}"))
-        .with_context(|| format!("failed to parse sierra program: {path}"))?;
+    let sierra_program = serde_json::from_str::<VersionedProgram>(
+        &fs::read_to_string(path.clone())
+            .with_context(|| format!("failed to read Sierra file: {path}"))?,
+    )
+    .map_err(|e| anyhow!("{e}"))
```
Comment (maciektr) [inline extensions/scarb-cairo-run/src/main.rs:69]: Do we need this line? It was added because of the way that lalrpop has constructed error structs.

File: extensions/scarb-cairo-run/src/main.rs [hunk 2]
```diff
@@ -61,13 +62,13 @@ fn main() -> Result<()> {
...
+    .with_context(|| format!("failed to deserialize sierra versioned program: {path}"))?
```
Comment (mkaput) [inline extensions/scarb-cairo-run/src/main.rs:70]:
```suggestion
    .with_context(|| format!("failed to deserialize Sierra program: {path}"))?
```

that was leaking implementation details to the user, also casing

File: extensions/scarb-cairo-run/src/main.rs (hunk 1 above)
Comment (mkaput) [reply extensions/scarb-cairo-run/src/main.rs:69]:
```suggestion
```

File: scarb/tests/build.rs [hunk 3]
```diff
@@ -528,10 +528,16 @@ fn sierra_replace_ids() {
         .assert()
         .success();
 
-    t.child("target/dev/hello.sierra")
-        .assert(predicates::str::contains(
-            "hello::example@0() -> (felt252);",
-        ));
+    t.child("target/dev/hello.sierra").assert(
```
Comment (mkaput) [inline scarb/tests/build.rs:531]: We need to discuss the extension with Cairo guys and ensure everybody likes it. Can you please start the conversation on our shared channel @MrDenkoV?

File: scarb/src/compiler/compilers/lib.rs [hunk 4]
```diff
@@ -60,11 +61,15 @@ impl Compiler for LibCompiler {
 
         if props.sierra {
```
Comment (mkaput) [inline scarb/src/compiler/compilers/lib.rs:62]: We need to give users an ability to generate textual representation, in case some tooling relies on it. Let's do that in separate task, that should be quite high priority, becase we need to deliver it in the same release as this PR: #634

File: scarb/src/compiler/compilers/lib.rs [hunk 5]
```diff
@@ -60,11 +61,15 @@ impl Compiler for LibCompiler {
...
+                format!("{}.sierra.json", unit.target().name),
                 "output file",
                 ws.config(),
             )?;
-            file.write_all(sierra_program.to_string().as_bytes())?;
+            file.write_all(
+                serde_json::to_string_pretty(&VersionedProgram::from((*sierra_program).clone()))
+                    .context("failed to serialize sierra versioned program")?
+                    .as_bytes(),
+            )?;
```
Comment (mkaput) [inline scarb/src/compiler/compilers/lib.rs:72]: 1. It's ineffective to generate text to string and then write it to file. You can do that in one operation: https://docs.rs/serde_json/latest/serde_json/fn.to_writer.html
2. `starknet-contract` generates minified JSON. Let's do that here too.

File: extensions/scarb-cairo-run/src/main.rs [hunk 6]
```diff
@@ -61,13 +62,15 @@ fn main() -> Result<()> {
...
-        .with_context(|| format!("failed to parse sierra program: {path}"))?;
+    let sierra_program = serde_json::from_str::<VersionedProgram>(
+        &fs::read_to_string(path.clone())
+            .with_context(|| format!("failed to read Sierra file: {path}"))?,
+    )
+    .with_context(|| format!("failed to deserialize Sierra program: {path}"))?
+    .into_v1()
+    .with_context(|| {
+        format!("failed to parse file into program - incorrect Sierra version: {path}")
+    })?;
```
Comment (mkaput) [inline extensions/scarb-cairo-run/src/main.rs:73]: 1. Using `-` to denote causability in errors seems inconsistent, given we have nice contexts mechanism in Anyhow
2. I think we can rely on `into_v1` function giving us the cause message, so maybe let's just do this:

```suggestion
    .into_v1()
    .with_context(|| format!("failed to load Sierra program: {path}"))?;
```

File: Cargo.toml [hunk 7]
```diff
@@ -28,18 +28,18 @@ repository = "https://github.com/software-mansion/scarb"
...
+cairo-lang-filesystem = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-formatter = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-language-server = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-runner = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-semantic = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-sierra = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-sierra-to-casm = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-starknet = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-test-runner = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" }
+cairo-lang-utils = { git = "https://github.com/starkware-libs/cairo", rev = "98af2f8" , features = ["env_logger"] }
```
Comment (mkaput) [inline Cargo.toml:42]: please extract Cairo update to a separate commit

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — COMMENTED** [review]
Looks good

**Review by mkaput — APPROVED** [review]
